torch/csrc/distributed/c10d/CUDASymmetricMemoryOps.cu

```diff
@@ -484,6 +489,63 @@ at::Tensor two_shot_all_reduce_(
   return input;
 }
 
+at::Tensor memset32_(
+    at::Tensor& input,
+    int64_t offset,
+    int64_t val,
+    int64_t count) {
+#if !defined(USE_ROCM) && defined(PYTORCH_C10_DRIVER_API_SUPPORTED)
+  TORCH_CHECK(
+      input.dim() == 1 && input.is_contiguous() &&
+          input.scalar_type() == c10::ScalarType::UInt32,
+      "CUDASymmetricMemoryUtils::memset32: input must be "
+      "a flat, contiguous uint32 tensor.");
+
+  TORCH_CHECK(
+      offset > 0 && count > 0,
+      "CUDASymmetricMemoryUtils::memset32: "
+      "offset and count must be a positive integers.")
+
+  TORCH_CHECK(
+      val >= 0 &&
+          static_cast<size_t>(val) <= std::numeric_limits<uint32_t>::max(),
+      "CUDASymmetricMemoryUtils::memset32: "
+      "val must be in the range of [0, 4294967295] (uint32_t).")
+
+  auto element_size = c10::elementSize(input.scalar_type());
+  TORCH_CHECK(
+      offset + count <= input.numel(),
+      "CUDASymmetricMemoryUtils::memset32: offset + count (",
+      offset + count,
+      ") exceeded the numel of the input (",
+      input.numel(),
+      ")");
+
+  auto driver_api = c10::cuda::DriverAPI::get();
+  CUcontext pctx, prev;
+
+  c10::cuda::CUDAGuard guard(input.device());
+  C10_CUDA_DRIVER_CHECK(driver_api->cuDevicePrimaryCtxRetain_(
+      &pctx, (CUdevice)input.device().index()));
+
+  C10_CUDA_DRIVER_CHECK(driver_api->cuCtxGetCurrent_(&prev));
```

Comment:
The [CUDA driver API docs](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__PRIMARY__CTX.html) say that:

> The primary context is unique per device and shared with the CUDA runtime API. These functions allow integration with other libraries using CUDA.

So I'm really puzzled: what context is this one that was set as current? I guess three options:
- it's the same as the primary one, in which case this code is not needed (but this can't be, or else you wouldn't have written it)
- it's a nullptr context, in which case to me this seems like there's some missing initialization happening (e.g., perhaps we're in a new/fresh thread that didn't issue any CUDA commands yet?)
- it's an actual non-primary CUDA context, but then I wonder: who created it?!

I think it would be very interesting to understand what is going on here.

Reply:
Additional references from the [CUDA runtime API docs](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DRIVER.html):

Primary

>  #### Contexts
> 
> There exists a one to one relationship between CUDA devices in the CUDA Runtime API and [CUcontext](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__TYPES.html#group__CUDA__TYPES_1gf9f5bd81658f866613785b3a0bb7d7d9) s in the CUDA Driver API within a process. The specific context which the CUDA Runtime API uses for a device is called the device's primary context. From the perspective of the CUDA Runtime API, a device and its primary context are synonymous.
> 
> #### Context Interoperability
> 
> Note that the use of multiple [CUcontext](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__TYPES.html#group__CUDA__TYPES_1gf9f5bd81658f866613785b3a0bb7d7d9) s per device within a single process will substantially degrade performance and is strongly discouraged. Instead, it is highly recommended that the implicit one-to-one device-to-context mapping for the process provided by the CUDA Runtime API be used.
> 
> If a non-primary [CUcontext](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__TYPES.html#group__CUDA__TYPES_1gf9f5bd81658f866613785b3a0bb7d7d9) created by the CUDA Driver API is current to a thread then the CUDA Runtime API calls to that thread will operate on that [CUcontext](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__TYPES.html#group__CUDA__TYPES_1gf9f5bd81658f866613785b3a0bb7d7d9), with some exceptions listed below. Interoperability between data types is discussed in the following sections.
> 
> The function [cudaPointerGetAttributes()](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__UNIFIED.html#group__CUDART__UNIFIED_1gd89830e17d399c064a2f3c3fa8bb4390) will return the error [cudaErrorIncompatibleDriverContext](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html#group__CUDART__TYPES_1gg3f51e3575c2178246db0a94a430e0038bfc1b14096f31fc7b43c08397af90856) if the pointer being queried was allocated by a non-primary context. The function [cudaDeviceEnablePeerAccess()](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__PEER.html#group__CUDART__PEER_1g2b0adabf90db37e5cfddc92cbb2589f3) and the rest of the peer access API may not be called when a non-primary [CUcontext](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__TYPES.html#group__CUDA__TYPES_1gf9f5bd81658f866613785b3a0bb7d7d9) is current. To use the pointer query and peer access APIs with a context created using the CUDA Driver API, it is necessary that the CUDA Driver API be used to access these features.
> 
> All CUDA Runtime API state (e.g, global variables' addresses and values) travels with its underlying [CUcontext](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__TYPES.html#group__CUDA__TYPES_1gf9f5bd81658f866613785b3a0bb7d7d9). In particular, if a [CUcontext](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__TYPES.html#group__CUDA__TYPES_1gf9f5bd81658f866613785b3a0bb7d7d9) is moved from one thread to another then all CUDA Runtime API state will move to that thread as well.
> 
> #### Initialization and Tear-Down
> 
> CUDA Runtime API calls operate on the CUDA Driver API [CUcontext](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__TYPES.html#group__CUDA__TYPES_1gf9f5bd81658f866613785b3a0bb7d7d9) which is current to to the calling host thread.
> 
> The function [cudaInitDevice()](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1gac04a5d82168676b20121ca870919419) ensures that the primary context is initialized for the requested device but does not make it current to the calling thread.
> 
> The function [cudaSetDevice()](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g159587909ffa0791bbe4b40187a4c6bb) initializes the primary context for the specified device and makes it current to the calling thread by calling [cuCtxSetCurrent()](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__CTX.html#group__CUDA__CTX_1gbe562ee6258b4fcc272ca6478ca2a2f7).
> 
> The CUDA Runtime API will automatically initialize the primary context for a device at the first CUDA Runtime API call which requires an active context. If no [CUcontext](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__TYPES.html#group__CUDA__TYPES_1gf9f5bd81658f866613785b3a0bb7d7d9) is current to the calling thread when a CUDA Runtime API call which requires an active context is made, then the primary context for a device will be selected, made current to the calling thread, and initialized.
> 
> The context which the CUDA Runtime API initializes will be initialized using the parameters specified by the CUDA Runtime API functions [cudaSetDeviceFlags()](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g69e73c7dda3fc05306ae7c811a690fac), [cudaD3D9SetDirect3DDevice()](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__D3D9.html#group__CUDART__D3D9_1g0733a6f0bdc8aca6c6812da66c6d814a), [cudaD3D10SetDirect3DDevice()](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__D3D10__DEPRECATED.html#group__CUDART__D3D10__DEPRECATED_1g7c942225222284fbfc3794a4ed6edf98), [cudaD3D11SetDirect3DDevice()](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__D3D11__DEPRECATED.html#group__CUDART__D3D11__DEPRECATED_1g533524c773e7ca4ef57c0698b157d670), [cudaGLSetGLDevice()](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__OPENGL__DEPRECATED.html#group__CUDART__OPENGL__DEPRECATED_1g88fddf479c2b597af49fd839bb42345e), and [cudaVDPAUSetVDPAUDevice()](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__VDPAU.html#group__CUDART__VDPAU_1g2a9ccfe8e6c87fe694bafd8cfcc67586). Note that these functions will fail with [cudaErrorSetOnActiveProcess](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html#group__CUDART__TYPES_1gg3f51e3575c2178246db0a94a430e0038a906ddf08d574274bf7334adb1497550) if they are called when the primary context for the specified device has already been initialized, except for [cudaSetDeviceFlags()](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g69e73c7dda3fc05306ae7c811a690fac) which will simply overwrite the previous settings.
> 
> Primary contexts will remain active until they are explicitly deinitialized using [cudaDeviceReset()](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1gef69dd5c6d0206c2b8d099abac61f217). The function [cudaDeviceReset()](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1gef69dd5c6d0206c2b8d099abac61f217) will deinitialize the primary context for the calling thread's current device immediately. The context will remain current to all of the threads that it was current to. The next CUDA Runtime API call on any thread which requires an active context will trigger the reinitialization of that device's primary context.
> 
> Note that primary contexts are shared resources. It is recommended that the primary context not be reset except just before exit or to recover from an unspecified launch failure.

Reply:
@lw You are absolutely right! I removed the calls to `cuDevicePrimaryCtxRetain`, `cuCtxGetCurrent`, and `cuCtxSetCurrent`, and everything still works. This part of the documentation also confirms it:

> The function [cudaSetDevice()](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g159587909ffa0791bbe4b40187a4c6bb) initializes the primary context for the specified device and makes it current to the calling thread by calling [cuCtxSetCurrent()](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__CTX.html#group__CUDA__CTX_1gbe562ee6258b4fcc272ca6478ca2a2f7).

I think this is how I reached the wrong conclusion that "you need to set the context for runtime and driver functions separately." Initially, I mistakenly used `cuMemsetD32` (the non-async version) and encountered an "invalid context" error. This made me wonder if I had to set the context for driver functions separately. As I was testing the theory, I noticed that I was calling the non-async version, so I switched to `cuMemsetD32Async`. Things worked afterward, and I attributed it to the wrong thing.

I tested again with `cuMemsetD32`. Even with the calls to `cuDevicePrimaryCtxRetain`, `cuCtxGetCurrent`, and `cuCtxSetCurrent`, I still get the "invalid context" error. I'm not sure if this is related to some "default stream semantics" caveats, but for now, it's irrelevant for the PR.

Thanks a lot! That was a great catch!